operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,411 @@
+func (c *cebManagerFcfs) RemoveCepFromCache(cep *cilium_v2.CoreCiliumEndpoint) {
+
+	cepName := getCepNameFromCCEP(cep)
+	log.WithFields(logrus.Fields{
+		"cep-name": cepName,
+	}).Debug("Remove CEP from local cache")
+
+	// Check in local cache, if a given cep is already batched in one of the ceb.
+	// and if exists, delete cep from ceb.
+	if cebName, exists := c.cacheCepMap.get(cepName); exists {
+		var ceb *ceBatch
+		if ceb, exists = c.cacheCeb[cebName]; !exists {
+			log.Info("Valid cep Mapping, but no matching CEP in CEB")
+			return
+		}
+		// Check for valid ceb object
+		if ceb == nil || ceb.ceb == nil {
+			log.Info("Valid cep Mapping, but not valid CEB")
+			return
+		}
+
+		ceb.backendMutex.Lock()
+		defer ceb.backendMutex.Unlock()
+		for i, ep := range ceb.ceb.Endpoints {
+			if getCepNameFromCCEP(&ep) == getCepNameFromCCEP(cep) {
+				// Insert deleted CoreCEP in removedCeps
+				ceb.removedCeps[getCepNameFromCCEP(&ep)] = struct{}{}
+				ceb.ceb.Endpoints =
+					append(ceb.ceb.Endpoints[:i],
+						ceb.ceb.Endpoints[i+1:]...)
+				break
+			}
+		}
+		log.WithFields(logrus.Fields{
+			"ceb-name":  cebName,
+			"cep-name":  cepName,
+			"cep-count": len(ceb.ceb.Endpoints),
+		}).Debug("Removed cep from ceb")
+
+		c.queue.Add(cebName)
+	}
+
+	return
+}
+
+// Return the total number of CEPs in cluster
+func (c *cebManagerFcfs) getTotalCepCount() int {
+	cnt := 0
+	for _, ceb := range c.cacheCeb {
+		ceb.backendMutex.RLock()
+		cnt += len(ceb.ceb.Endpoints)
+		ceb.backendMutex.RUnlock()
+	}
+	return cnt
+}
+
+// Return the total number of CEPs in the ceb
+func (c *cebManagerFcfs) getCepCountInCeb(cebName string) (cnt int) {
+	if ceb, ok := c.cacheCeb[cebName]; ok {
+		ceb.backendMutex.RLock()
+		cnt = len(ceb.ceb.Endpoints)
+		ceb.backendMutex.RUnlock()
+	}
+	return
+}
+
+// Return the total count of CEB
+func (c *cebManagerFcfs) getCebCount() int {
+	return len(c.cacheCeb)
+}
+
+// Return list of Core CEPs
+func (c *cebManagerFcfs) getAllCeps() map[string]*cilium_v2.CoreCiliumEndpoint {
+	ceps := make(map[string]*cilium_v2.CoreCiliumEndpoint)
+	for _, ceb := range c.cacheCeb {
+		ceb.backendMutex.RLock()
+		for _, cep := range ceb.ceb.Endpoints {
+			ceps[getCepNameFromCCEP(&cep)] = cep.DeepCopy()
+		}
+		ceb.backendMutex.RUnlock()
+	}
+
+	return ceps
+}
+
+// Return the list of removed Core CEPs
+func (c *cebManagerFcfs) getRemovedCeps(cebName string) map[string]struct{} {
+	cepNames := make(map[string]struct{})
+	if ceb, _ := c.cacheCeb[cebName]; ceb != nil {
+		ceb.backendMutex.RLock()
+		for cepName := range ceb.removedCeps {
+			cepNames[cepName] = struct{}{}
+		}
+		ceb.backendMutex.RUnlock()
+	}
+
+	return cepNames
+}
+
+// After successful sync with api-server, delete removed ceps in a CEB.
+// If no more CEPs are packed in CEB, Delete the CEB in next DeleteSYNC.
+func (c *cebManagerFcfs) clearRemovedCeps(cebName string, remCeps map[string]struct{}) {
+
+	var ok bool
+	var ceb *ceBatch
+	// Check if CEB exists in local cache
```

Comment:
Can you comment here when this would happen?

Reply:
Serves 2 purposes, get `ceb` object mapped to cebName and bail out if there is no such map exists.

Reply:
These are the guardrails to prevent invalid accesses and eventual crash. In multi-threaded environments, there is no guarantee of events ordering. what happens if request for `CEB` delete request succeeds, and we deleted cache, but for the same CEB `UPDATE` got delayed and calls `clearRemovedCeps`

Reply:
Why would you have a `DELETE` and an `UPDATE` for a CEB in the same time?

Yes mutlithreaded env can have many event orderings, but I do not think blindly throwing guardrails here and there is a good practice, we should understand whether a condition could happen, if it does, then handle it, otherwise we should not leave dead code in production.

In your code, when you call `clearRemovedCeps`, you are in the middle of processing a CEB, it should always be in the cache, if not, that is a bug.

Reply:
I strongly believe, checking for valid  `object` or `pointer` is the right thing to do and a good practice. I'll leave as it is, unless other reviewers feels its unnecessary.

Reply:
Maybe throw a warning here? It should never happen, if it does, then something is not right and it'll be good to have a trace in the log.